src/mame/yamaha/ymdx9.cpp

```diff
@@ -0,0 +1,449 @@
+
+	// Configure the ADC. The clock speed here is a guess.
+	M58990(config, m_adc, 8_MHz_XTAL / 16);
+
+	// ADC source 4 is the battery voltage. Set this input to always read 0x80.
+	// If the read value is below 0x6F, the firmware considers this a low battery voltage.
+	m_adc->in_callback<4>().set_constant(0x80);
+
+	// Configure MIDI.
+	auto &midiclock(CLOCK(config, "midiclock", 500_kHz_XTAL));
+	midiclock.signal_handler().set(FUNC(yamaha_dx9_state::midiclock_w));
+
+	MIDI_PORT(config, "mdin", midiin_slot, "midiin").rxd_handler().set(FUNC(yamaha_dx9_state::midi_r));
+
+	auto &mdout(MIDI_PORT(config, "mdout", midiout_slot, "midiout"));
+	m_maincpu->out_ser_tx_cb().set(mdout, FUNC(midi_port_device::write_txd));
+
+	// Configure the LCD screen.
+	screen_device &lcd_screen(SCREEN(config, "lcd_screen", SCREEN_TYPE_LCD));
+	lcd_screen.set_refresh_hz(60);
+	lcd_screen.set_vblank_time(ATTOSECONDS_IN_USEC(2500)); /* not accurate */
+	lcd_screen.set_screen_update("lcdc", FUNC(hd44780_device::screen_update));
+	lcd_screen.set_size(6*16+1, 10*2+1);
+	lcd_screen.set_visarea_full();
+	lcd_screen.set_palette("palette");
+
+	PALETTE(config, "palette", FUNC(yamaha_dx9_state::palette_init), 3);
+
+	hd44780_device &lcdc(HD44780(config, "lcdc", 0));
+	lcdc.set_lcd_size(2, 16);
+	lcdc.set_pixel_update_cb(FUNC(yamaha_dx9_state::lcd_pixel_update));
+
+	config.set_default_layout(layout_dx9);
+}
+
+
+/**
+ * yamaha_dx9_state::unmapped_r
+ */
+uint8_t yamaha_dx9_state::unmapped_r(offs_t offset)
+{
+	return 0xFF;
+}
+
+
+/**
+ * yamaha_dx9_state::key_switch_scan_driver_r
+ */
+uint8_t yamaha_dx9_state::key_switch_scan_driver_r(offs_t offset)
+{
+	/**
+	 * @brief The value to be returned.
+	 * This is implemented as a separate variable to allow for much easier debugging, as
+	 * opposed to just instantly returning the value read from the input lines.
+	 */
+	uint8_t value = 0;
+
+	// Note: Input line 4 covers the keyboard circuit, which is not implemented here.
+	// When the keyboard state is read, the default value of 0 will be returned.
+	if(m_key_switch_scan_driver_select == 0)
+	{
+		// Input line 0 covers the 'main' front-panel switches.
+		value = m_key_switch_input_driver_line_0->read();
+	} else if(m_key_switch_scan_driver_select == 1)
+	{
+		// Input line 1 covers the numeric front-panel switches 1 through 8.
+		value = m_key_switch_input_driver_line_1->read();
+	} else if(m_key_switch_scan_driver_select == 2)
+	{
+		// Input line 2 covers the numeric front-panel switches 9 though 16.
+		value = m_key_switch_input_driver_line_2->read();
+	} else if(m_key_switch_scan_driver_select == 3)
+	{
+		// Input line 3 covers the numeric front-panel switches 17 though 20.
+		// The Portamento, and Sustain pedals are mapped to bits 6, and 7 respectively.
+		value = m_key_switch_input_driver_line_3->read();
+	}
+
+	return value;
+}
+
+
+/**
+ * yamaha_dx9_state::unmapped_w
+ */
+void yamaha_dx9_state::unmapped_w(offs_t address, uint8_t data) {}
```

Comment:
Not needed. If you want to silence writes to a memory location you can use `nopw()`.